connection_manager.go

```diff
@@ -0,0 +1,393 @@
+// options
+func NewConnectionManager(opts ConnectionManagerOptions) (ConnectionManager, error) {
+	if opts == nil {
+		opts = NewConnectionManagerOptions()
+	}
+
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	drivers := make(map[string]Driver, len(opts.GetDrivers()))
+	for _, driver := range opts.GetDrivers() {
+		drivers[driver.Type().Name()] = driver
+	}
+
+	return &connectionManager{
+		log:      opts.GetLogger(),
+		conns:    make(map[string]Connection),
+		pending:  make(map[string]<-chan struct{}),
+		watches:  make(map[string]ClusterWatch),
+		drivers:  drivers,
+		provider: opts.GetProvider(),
+	}, nil
+}
+
+type connectionManager struct {
+	sync.RWMutex
+
+	log      xlog.Logger
+	conns    map[string]Connection
+	pending  map[string]<-chan struct{}
+	watches  map[string]ClusterWatch
+	drivers  map[string]Driver
+	provider ClusterMappingProvider
+	closed   bool
+	wg       sync.WaitGroup
+}
+
+// GetConnection retrieves a connection to the given database and cluster
+func (cm *connectionManager) GetConnection(database, cluster string) (Connection, error) {
+	cm.log.Debugf("retrieving connection for %s:%s", database, cluster)
+	for {
+		if conn, err := cm.getExistingConn(database, cluster); conn != nil || err != nil {
+			return conn, err
+		}
+
+		if conn, err := cm.tryOpenConn(database, cluster); conn != nil || err != nil {
+			return conn, err
+		}
+	}
+}
+
+// Close closes the connection manager
+func (cm *connectionManager) Close() error {
+	cm.Lock()
+	if cm.closed {
+		cm.Unlock()
+		return nil
+	}
+
+	cm.log.Infof("closing connection manager")
+
+	// Mark ourselves as closed
+	cm.closed = true
+
+	// Close all watches.  The watch goroutine will take care of cleaning up
+	// connections and the watch map entry
+	for _, cw := range cm.watches {
+		cw.Close()
+	}
+
+	// Wait for watch goroutines to exit
+	cm.Unlock()
+	cm.wg.Wait()
+	return nil
+}
+
+// getExistingConn attmepts to return an existing open connection to the database cluster
+func (cm *connectionManager) getExistingConn(database, cluster string) (Connection, error) {
+	key := fmtConnKey(database, cluster)
+
+	cm.RLock()
+	defer cm.RUnlock()
+
+	// If we're closed, bail out
+	if cm.closed {
+		return nil, errClosed
+	}
+
+	// If there is already a connection, return it
+	if c := cm.conns[key]; c != nil {
+		return c, nil
+	}
+
+	// No connection, no error - try to create it
+	return nil, nil
+}
+
+// tryOpenConn tries to create a new connection to the database cluster
+func (cm *connectionManager) tryOpenConn(database, cluster string) (Connection, error) {
+	key := fmtConnKey(database, cluster)
+
+	cm.log.Infof("attempting to create connection to %s:%s", database, cluster)
+
+	cm.Lock()
+
+	// If someone closed us while we were upgrading the lock, bail out
+	if cm.closed {
+		return nil, errClosed
+	}
+
+	// If someone else created it while we were upgrading the lock, return the
+	// new connection
+	if c := cm.conns[key]; c != nil {
+		cm.Unlock()
```

Comment:
We do - see the `GetConnection` code which is the main line for retrieving a connection .  You only enter `tryOpenConn` if there is no existing connection obtainable under a read lock, and you are going to try to open one.